### PR TITLE
Improvements for the OAI-PMH Server.

### PR DESCRIPTION
**Pull request by EreMaijala**
- Use cursorMark instead of deep paging with Solr
- Make it possible to override the Server class
- Ignore records that cannot be returned in the requested format

cursorMark is the most important one, since deep paging cannot be used with large and/or sharded indexes. I've tried to make it generic enough that the implementation could be used with other sources too.

I believe that ignoring records that cannot be returned is the proper way to handle repositories with several metadata formats.

TODO:
- [X] Make sure that cursorMark method takes into account the Solr listeners that may add other filters etc.

### CODE REVIEW

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for this! I've found a few issues that I think need to be addressed based on a code review... once you've had a chance to look at those, let me know, and I'll proceed with some hands-on testing to confirm that things are working as expected.

Whatever happens, I think we'll want to hold off on merging this until we begin work on 6.0, since this is a backward-breaking change... but I don't think that's very far off, with the Summit coming up in a couple of weeks, and not much on the slate for 5.1.

File: module/VuFind/src/VuFind/Search/Base/Results.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Search/Base/Results.php]: Okay, I'll revert to Solr-specific implementation.

File: module/VuFind/src/VuFind/OAI/Server.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/OAI/Server.php]: I'll verify and fix as needed.

**Comment by EreMaijala** [comment]
@demiankatz  Thanks for the review, all observations were spot on. They should now be addressed.

File: module/VuFind/src/VuFind/OAI/Server.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/OAI/Server.php]: If we continue here, is the fact that we're not incrementing $currentCursor as a result going to throw anything off? Seems like we might get duplicate or unexpected results....

**Comment by demiankatz** [comment]
@EreMaijala, thanks for the fixes. I just added one new comment regarding the skipping of unsupported metadata records.

**Comment by EreMaijala** [comment]
The skipping should be fixed now. 

It would be neat if we had record type indexed and mapping that would tell which types are available in the requested format so that skipping could be avoided altogether.

**Comment by demiankatz** [comment]
Regarding indexing of record type, see #1260, which upgrades the recordtype field to record_format for consistency with the authority index (as suggested in my earlier comment). Hopefully we can merge this shortly after we begin work on the 6.0 release and can then take advantage of it here.

**Review by demiankatz — APPROVED** [review]

**Comment by demiankatz** [comment]
@EreMaijala, this seems to be working well now, so I have merged it. I realize there is some open discussion about making better use of the record_format field after #1260 is merged, but I think it's just as well to deal with that in a separate PR if/when the time comes to work on it.

**Pull request merged**
